File: elixir/2015/day01a/Program.exs
#!/usr/bin/env elixir

defmodule Program do
  def main(args) do
    case args do
      [filename] ->
        case process(filename) do
          {:ok, result} ->
            IO.puts("result = #{result}")
          {:error, reason} ->
            IO.puts(:stderr, "error while processing file `#{filename}': #{reason}")
            System.halt(1)
        end
      _ ->
        usage()
    end
  end

  defp usage do
    IO.puts(:stderr, "usage: Program.exs <input file>")
    System.halt(1)
  end

  defp process(filename) do
    case File.read(filename) do
      {:ok, content} ->
        floors =
          content
          |> String.graphemes()
          |> Enum.reduce(0, fn
            "(", acc -> acc + 1
            ")", acc -> acc - 1
            _, acc -> acc
          end)

        {:ok, floors}
      {:error, reason} ->
        {:error, reason}
    end
  end
end

Program.main(System.argv())

File: erlang/2015/day01a/Program.erl
#!/usr/bin/env escript

main(Args) ->
  case Args of
    [Filename] ->
      case process(Filename) of
        {ok, Result} ->
          io:format("result = ~p~n", [Result]);
        {error, Reason} ->
          io:format(standard_error, "error while processing file `~s': ~p~n", [Filename, Reason]),
          halt(1)
      end;
    _ ->
      usage()
  end.

usage() ->
  io:format(standard_error, "usage: Program.erl <input file>~n", []),
  halt(1).

process(Filename) ->
  case file:read_file(Filename) of
    {ok, Chars} ->
      Content = unicode:characters_to_list(Chars),
      Result = lists:foldl(
        fun
          ($(, Acc) -> Acc + 1;
          ($), Acc) -> Acc - 1;
          (_, Acc) -> Acc
        end,
        0,
        Content),
      {ok, Result};
    {error, Reason} ->
      {error, Reason}
  end.
